### PR TITLE
Added LineGraph generator class complete with unit test

### PR DESCRIPTION
**Pull request by nks1558**
Implemented LineGraph generator method which converts a given graph to its corresponding Line graph.

### CODE REVIEW

File: jgrapht-core/src/main/java/org/jgrapht/generate/LineGraphGenerator.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+            List<E> edgeList = new ArrayList<>(graph.edgeSet());
+
+            for (E edge1 : edgeList) {
+                for (E edge2 : edgeList) {
+                    if (edge1 == edge2) {
+                        continue;
+                    }
+
+                    if (graph.getEdgeTarget(edge1) == graph.getEdgeSource(edge2)) {
+                        target.addVertex((V) edge1);
```
Comment (jkinable) [inline jgrapht-core/src/main/java/org/jgrapht/generate/LineGraphGenerator.java]: You cannot do this. This results in unchecked cast warnings. See: https://github.com/jgrapht/jgrapht/wiki/Contributor-Guidelines 
You are not supposed to introduce new warnings. However, that's not really the problem. Try adding the following code to the end of your `testDirectedGraph()` method:
```
for(Integer v : target.vertexSet())
            System.out.println(v);
```
This will give you the following error: `java.lang.ClassCastException: org.jgrapht.graph.DefaultEdge cannot be cast to java.lang.Integer`

Comment (nks1558) [reply jgrapht-core/src/main/java/org/jgrapht/generate/LineGraphGenerator.java]: Hey Joris,
I will fix all the issues you just mentioned. 
But I am not clear about this one. What I make of a line graph is that its vertex set has "edges" as its vertices. So, when I find that a new edge is to be added in the line graph, I add the edge as a vertex i.e. 'target.addVertex((V) edge)' . Now for adding an edge as a vertex, this cast was required. Can you please suggest a workaround.
Also please correct me if my implication of a line graph is correct or not.
I will fix the remaining issues and push the modified code as soon as possible.

Comment (nks1558) [reply jgrapht-core/src/main/java/org/jgrapht/generate/LineGraphGenerator.java]: Hey Joris,
I will fix all the issues you just mentioned.
But I am not clear about this one. What I make of a line graph is that its vertex set has "edges" as its vertices. So, when I find that a new edge is to be added in the line graph, I add the edge as a vertex i.e. `target.addVertex((V) edge)` . Now for adding an edge as a vertex, this cast was required. Can you please suggest a workaround.
Also please correct me if my implication of a line graph is correct or not.
I will fix the remaining issues and push the modified code as soon as possible.

**Comment by jsichi** [comment]
I don't think a generator is the right packaging for this.  Generators are supposed to make up graphs out of thin air, whereas this is actually a transformation of an existing graph.  The right place is probably a new subpackage org.jgrapht.alg.dual.  @jkinable what's your opinion?

File: jgrapht-core/src/test/java/org/jgrapht/generate/LineGraphGeneratorTest.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        g.addEdge(4,1);
+        g.addEdge(4,3);
+        g.addEdge(1,3);
+
+        LineGraphGenerator<Integer, DefaultEdge> lgg = new LineGraphGenerator<>(g);
+        Graph<Integer, DefaultEdge> target = new SimpleGraph<>(DefaultEdge.class);
+        lgg.generateGraph(target, null);
+
+        assertTrue(target.vertexSet().equals(g.edgeSet()));
+        assertEquals(9, target.edgeSet().size());
```
[the remark this answers is not in the document]
Comment (nks1558) [reply jgrapht-core/src/test/java/org/jgrapht/generate/LineGraphGeneratorTest.java]: I think I have missed the verification of the edges of the new Line graph. But I am facing an issue here. The edges of the line graph have `DefaultEdge` as its source and target vertex. How do I check an edge between 2 edges? Should I construct the Line graph manually and then compare it with the output?

File: jgrapht-core/src/main/java/org/jgrapht/generate/LineGraphGenerator.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+
+    @Override
+    public void generateGraph(Graph<V, E> target, Map<String, V> resultMap)
+    {
+        if (graph.getType().isDirected()) {
+            List<E> edgeList = new ArrayList<>(graph.edgeSet());
+
+            for (E edge1 : edgeList) {
```
[the remark this answers is not in the document]
Comment (nks1558) [reply jgrapht-core/src/main/java/org/jgrapht/generate/LineGraphGenerator.java]: Thanks for this modification.
It will become O(V+E), right?

File: jgrapht-core/src/test/java/org/jgrapht/generate/LineGraphGeneratorTest.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        g.addEdge(3,4);
+        g.addEdge(4,3);
+
+        LineGraphGenerator<Integer, DefaultEdge> lgg = new LineGraphGenerator<>(g);
+        Graph<Integer, DefaultEdge> target = new SimpleDirectedGraph<>(DefaultEdge.class);
+        lgg.generateGraph(target, null);
+
+        assertTrue(target.vertexSet().equals(g.edgeSet()));
+        assertEquals(12, target.edgeSet().size());
+    }
```
[the remark this answers is not in the document]
Comment (nks1558) [reply jgrapht-core/src/test/java/org/jgrapht/generate/LineGraphGeneratorTest.java]: I have added the star graph special case.
Bu the `GraphTests` class does not contain tests for graphs like `Square Graph` , `Triangle Graph` , `Cycle Graph` etc. So, I am not able to test these corner cases directly. Can these tests be added in the `GraphTests` class?

File: jgrapht-core/src/main/java/org/jgrapht/generate/LineGraphGenerator.java [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+import org.jgrapht.Graph;
+
+import java.util.*;
+
+/**
+ * Generator which produces the
+ * <a href="http://mathworld.wolfram.com/LineGraph.html">line graph</a> of a given input
+ * graph.
+ * The line graph of an undirected graph G is another graph L(G) that represents the
```
[the remark this answers is not in the document]
Comment (nks1558) [reply jgrapht-core/src/main/java/org/jgrapht/generate/LineGraphGenerator.java]: Done

File: jgrapht-core/src/test/java/org/jgrapht/generate/LineGraphGeneratorTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (jkinable) [reply jgrapht-core/src/test/java/org/jgrapht/generate/LineGraphGeneratorTest.java]: yes

File: jgrapht-core/src/test/java/org/jgrapht/generate/LineGraphGeneratorTest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (jkinable) [reply jgrapht-core/src/test/java/org/jgrapht/generate/LineGraphGeneratorTest.java]: no not needed. Just add some tests in which you manually check the result

File: jgrapht-core/src/main/java/org/jgrapht/generate/LineGraphGenerator.java (hunk 1 above)
Comment (jkinable) [reply jgrapht-core/src/main/java/org/jgrapht/generate/LineGraphGenerator.java]: Your interpretation is correct. Give it a try yourself first.

**Comment by d-michail** [comment]
There is a major flaw in this implementation. 

Briefly, the generics parameters of (a) the input graph and (b) the output graph should not be the same. This is the reason that you perform an unsafe cast which will result in `ClassCastException` during runtime. As @jsichi suggested mentioned, the generator interface is not the right way to go. 

I would suggest to first become familiar with our code base, read a little bit about Java generics (https://docs.oracle.com/javase/tutorial/java/generics/types.html) which are used everywhere in the library, and then dive into coding.

**Comment by nks1558** [comment]
Thankyou Michail for providing feedback. I will definitely learn the required concepts and will get familiar with your codebase. It would be very helpful if you can tell me in which package should I put this. I followed the implementation of `ComplementGraphGenerator` which produces the comlement of an input graph, so I thought LineGraphGenerator will also take in an input graph and return the corresponding Line graph.

**Comment by d-michail** [comment]
Maybe a package `org.jgrapht.alg.dual` but still the class must not implement the generator interface.

It should get as input a Graph<V,E1> and return a Graph<E1,E2> thus the class should have three generic parameters V, E1, E2.

**Comment by nks1558** [comment]
Hey michail,
I think I have made the required corrections(No unsafe cast is being done now). Now I will put the implementation in the suggested `org.jgrapht.alg.dual` package. I would like to know the naming conventions for this new package. Can I retain the class name as `LineGraphGenerator`?
Any new suggestion.
Thanks,
Nikhil

**Comment by jsichi** [comment]
Regarding naming, how about LineGraphConstructor?

(Later we will want a LineGraphRecognizer which can also produce the root graph.)

**Comment by jkinable** [comment]
@jsichi I would prefer to rename the package to 'org.jgrapht.alg.transform'. Dual has a specific meaning in math, and also in graph theory, e.g. 'dual graph'. A line graph _transformation_ is applied to an input graph, so I think a transform package could make sense. 
Also, wrt the class name, we could use our naming convention 'AsLineGraph' (e.g. similar to AsUndirectedGraph). What do you think?

**Comment by d-michail** [comment]
Regarding the package, I also prefer the transform. The `AsSomething` convention is more like a view over something, not sure whether is the right fit here.

**Comment by jsichi** [comment]
I'm fine with transform for the package.  (I suggested dual because per wikipedia and mathworld, the line graph is also referred to as the edge-to-vertex dual, but we can keep dual reserved for the more common polyhedral meaning).

I agree that we should avoid "As" since we already use it for live views only.  I'm fine with "To" per the suggestion from @AlexandruValeanu since it has the virtue of being shorter than "Constructor" or "Transform".  Whatever we decide on will be established as the convention for future transformation algorithms.

File: jgrapht-core/src/main/java/org/jgrapht/alg/dual/LineGraphConstructor.java [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    {
+        this.graph = graph;
+    }
+
+    /**
+     * Constructs line graph of a given graph.
+     *
+     * @param target target graph
+     */
+    public void constructGraph(Graph<E1, E2> target)
```
Comment (jkinable) [inline jgrapht-core/src/main/java/org/jgrapht/alg/dual/LineGraphConstructor.java]: Are there any requirements on the target graph? If I recall correctly, the line graph of a directed graph is also directed. Consider using GraphTests.requireDirectedGraph(.). Maybe something similar for the undirected case

File: jgrapht-core/src/main/java/org/jgrapht/alg/dual/LineGraphConstructor.java [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+ * Generator which produces the
+ * <a href="http://mathworld.wolfram.com/LineGraph.html">line graph</a> of a given input
+ * graph.
+ * The line graph of an undirected graph $G$ is another graph $L(G)$ that represents the
+ * adjacencies between edges of $G$.
+ * The line graph of a directed graph $G$ is the directed graph $L(G)$ whose vertex set
+ * corresponds to the arc set of $G$ and having an arc directed from an edge $e_1$ to an
+ * edge $e_2$ if in $G$, the head of $e_1$ meets the tail of $e_2$
+ *
+ * The runtime complexity is $O(|V|+|E|)$.
```
Comment (jkinable) [inline jgrapht-core/src/main/java/org/jgrapht/alg/dual/LineGraphConstructor.java]: This runtime complexity does not seem to be correct.

File: jgrapht-core/src/test/java/org/jgrapht/alg/dual/LineGraphConstructorTest.java [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+
+        assertTrue(GraphTests.isComplete(target));
+    }
+
+    @Test
+    public void testUndirectedGraph()
+    {
+        Graph<Integer, DefaultEdge> g = new SimpleGraph<>(DefaultEdge.class);
+        Graphs.addAllVertices(g, Arrays.asList(1, 2, 3, 4, 5));
+        g.addEdge(1,2);
```
Comment (jkinable) [inline jgrapht-core/src/test/java/org/jgrapht/alg/dual/LineGraphConstructorTest.java]: If you sue DefaultEdge e1=g.addEdge(1,2); then below you can use expectedGraph.addEdge(e1, e2);

File: jgrapht-core/src/main/java/org/jgrapht/alg/dual/LineGraphConstructor.java [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+                    for(E1 edge2 : graph.outgoingEdgesOf(vertex)){
+                        target.addVertex(edge1);
+                        target.addVertex(edge2);
+                        target.addEdge(edge1, edge2);
+                    }
+                }
+            }
+        } else{ // undirected graph
+
+            for(E1 edge : graph.edgeSet()){
```
Comment (jkinable) [inline jgrapht-core/src/main/java/org/jgrapht/alg/dual/LineGraphConstructor.java]: Removing the recurring addVertex() calls.
It also seems this part of the implementation can be made more efficient. Not sure at the moment, would need to have another look.

**Comment by jkinable** [comment]
@AlexandruValeanu The 'ToLineGraph' seemed like a good idea to me, but after looking into this a bit more I think that 'LineGraphTransformer' or something similar might be more applicable, since at some point, we might want to add methods such as:
`isLineGraph(graph)` and getRootGraph(graph) which is the inverse operation of the line graph tranformation. `ToLineGraph` sounds like it can only do 1 thing.

**Comment by AlexandruValeanu** [comment]
@jkinable I have another idea but it involves more classes: have both `ToLineGraph` and `FromLineGraph`. This way everything that is specific to a (one-way) transformation gets its own class and the docs can mention the reverse transformation. Thus it is very clear what every class does (since it only does one thing). Both `To...` and `From...` can be placed in the same package.

`isLineGraph(graph)` should be in `GraphTests`, shouldn't it?

I also agree with yours as it keeps a smaller number of classes that can do multiple things.

**Comment by d-michail** [comment]
Not a bad idea. They could also implement some generic interface like Function<T,R> with T the input graph type and R the output graph type.

File: jgrapht-core/src/main/java/org/jgrapht/alg/dual/LineGraphConstructor.java [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+ * adjacencies between edges of $G$.
+ * The line graph of a directed graph $G$ is the directed graph $L(G)$ whose vertex set
+ * corresponds to the arc set of $G$ and having an arc directed from an edge $e_1$ to an
+ * edge $e_2$ if in $G$, the head of $e_1$ meets the tail of $e_2$
+ *
+ * The runtime complexity is $O(|V|+|E|)$.
+ *
+ * <p>
+ * More formally, let $G = (V, E)$ be a graph then its line graph $L(G)$ is such that
+ * - each vertex of $L(G)$ represents an edge of $G$; and
```
[the remark this answers is not in the document]
Comment (nks1558) [reply jgrapht-core/src/main/java/org/jgrapht/alg/dual/LineGraphConstructor.java]: Done.

File: jgrapht-core/src/main/java/org/jgrapht/alg/dual/LineGraphConstructor.java [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+ *   a common endpoint ("are incident") in $G$.
+ * <p>
+ *
+ * @author Nikhil Sharma
+ * @since June 2018
+ *
+ *
+ * @param <V> vertex type
+ * @param <E1> edge type
+ * @param <E2> edge type
```
[the remark this answers is not in the document]
Comment (nks1558) [reply jgrapht-core/src/main/java/org/jgrapht/alg/dual/LineGraphConstructor.java]: Done. Now using only E

File: jgrapht-core/src/main/java/org/jgrapht/alg/dual/LineGraphConstructor.java [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+    private final Graph<V, E1> graph;
+
+    /**
+     * Line Graph Constructor
+     *
+     * @param graph input graph
+     */
+    public LineGraphConstructor(Graph<V, E1> graph)
+    {
+        this.graph = graph;
```
[the remark this answers is not in the document]
Comment (nks1558) [reply jgrapht-core/src/main/java/org/jgrapht/alg/dual/LineGraphConstructor.java]: Done

File: jgrapht-core/src/main/java/org/jgrapht/alg/dual/LineGraphConstructor.java [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+     * @param target target graph
+     */
+    public void constructGraph(Graph<E1, E2> target)
+    {
+        if (graph.getType().isDirected()) {
+
+            for(V vertex : graph.vertexSet()){
+                for(E1 edge1 : graph.incomingEdgesOf(vertex)){
+                    for(E1 edge2 : graph.outgoingEdgesOf(vertex)){
+                        target.addVertex(edge1);
```
[the remark this answers is not in the document]
Comment (nks1558) [reply jgrapht-core/src/main/java/org/jgrapht/alg/dual/LineGraphConstructor.java]: Corrected.

File: jgrapht-core/src/main/java/org/jgrapht/alg/dual/LineGraphConstructor.java (hunk 6 above)
Comment (nks1558) [reply jgrapht-core/src/main/java/org/jgrapht/alg/dual/LineGraphConstructor.java]: Done

File: jgrapht-core/src/main/java/org/jgrapht/alg/dual/LineGraphConstructor.java (hunk 7 above)
Comment (nks1558) [reply jgrapht-core/src/main/java/org/jgrapht/alg/dual/LineGraphConstructor.java]: Changed to O(V). Please correct me if I am wrong.

File: jgrapht-core/src/test/java/org/jgrapht/alg/dual/LineGraphConstructorTest.java (hunk 8 above)
Comment (nks1558) [reply jgrapht-core/src/test/java/org/jgrapht/alg/dual/LineGraphConstructorTest.java]: Done

File: jgrapht-core/src/main/java/org/jgrapht/alg/dual/LineGraphConstructor.java (hunk 9 above)
Comment (nks1558) [reply jgrapht-core/src/main/java/org/jgrapht/alg/dual/LineGraphConstructor.java]: Removed recurring addVertex() calls and using `Graphs.addAllVertices(target, graph.edgeSet())`

**Comment by nks1558** [comment]
Hello all,
I think I have made the required changes but I am still not sure about the package and class names as there was a lot of discussion. If you can provide me with the final names, I will push the changes as soon as possible.

**Comment by nks1558** [comment]
I know my contributions are not really professional but this is my first attempt at open source. I would be very happy if you can provide me with feedback.

**Comment by jkinable** [comment]
@nks1558 can you go ahead and make those final changes? We are getting closer to the next release. It would be nice to support line graphs.
